### PR TITLE
Implementing Api construct

### LINKED ISSUE #22: Create an API construct
Make it easier to create APIs by creating a higher level construct. Need to automatically handle things like cold starts, etc.

### TODO
- [ ] easy way to enable CORS, if both cors and corsPreflight are defined, cors takes precedence
- [ ] enable access log by default, allow customizing log format
- [ ] enable xray by default for all lambdas
- [ ] enable AWS_NODEJS_CONNECTION_REUSE_ENABLED by default for all lambdas
- [ ] easy way to enable AWS_IAM authorization for all routes, allow override per route
- [ ] easy way to define common Lambda props (environment, policy, srcPath, etc) for all aroutes, allow override per route

### PR DESCRIPTION
**Pull request by fwang**
Close #22

### CODE REVIEW

File: packages/resources/src/Api.ts [hunk 1]
```diff
@@ -58,158 +74,183 @@ export interface RouteProps {
...
       // Get Lambda props
-      const lambdaProps = { ...defaultLambdaProps, ...routeProps.lambdaProps };
-      if ( ! lambdaProps.handler) {
-        throw new Error(`No Lambda handler defined for route ${routeName}`);
+      const lambdaProps = {
+        ...(defaultLambdaProps || {}),
+        ...routeProps.lambdaProps,
+      } as FunctionProps;
+      if (!lambdaProps.handler) {
+        throw new Error(`No Lambda handler defined for route ${routeKey}`);
```
Comment (fwang) [inline packages/resources/src/Api.ts:229]: copy

File: packages/resources/src/Api.ts [hunk 2]
```diff
@@ -58,158 +74,183 @@ export interface RouteProps {
...
 
       // Get authorization type
-      let authorizationType = routeProps.authorizationType || defaultAuthorizationType || 'NONE';
+      let authorizationType =
+        routeProps.authorizationType || defaultAuthorizationType || "NONE";
       authorizationType = authorizationType.toUpperCase();
-      if ( ! [ 'NONE', 'AWS_IAM' ].includes(authorizationType)) {
+      if (!["NONE", "AWS_IAM"].includes(authorizationType)) {
         throw new Error(
           `sst.Api does not support ${authorizationType} authorization type. Only 'NONE' and 'AWS_IAM' types are currently supported.`
```
Comment (fwang) [inline packages/resources/src/Api.ts:219]: copy

File: packages/resources/src/Api.ts [hunk 3]
```diff
@@ -58,158 +74,183 @@ export interface RouteProps {
...
-        apig.HttpMethod.PUT,
-      ].includes(method)) {
-        throw new Error(`Invalid method defined for route ${routeName}`);
+      const method = allowedMethods.find((per) => per === methodStr);
+      if (!method) {
+        throw new Error(`Invalid method defined for route ${routeKey}`);
       }
       if (path.length === 0) {
-        throw new Error(`Invalid path defined for route ${routeName}`);
+        throw new Error(`Invalid path defined for route ${routeKey}`);
```
Comment (fwang) [inline packages/resources/src/Api.ts:210]: copy

File: packages/resources/src/Api.ts [hunk 4]
```diff
@@ -58,158 +74,183 @@ export interface RouteProps {
...
-        apig.HttpMethod.HEAD,
-        apig.HttpMethod.OPTIONS,
-        apig.HttpMethod.PATCH,
-        apig.HttpMethod.POST,
-        apig.HttpMethod.PUT,
-      ].includes(method)) {
-        throw new Error(`Invalid method defined for route ${routeName}`);
+      const method = allowedMethods.find((per) => per === methodStr);
+      if (!method) {
+        throw new Error(`Invalid method defined for route ${routeKey}`);
```
Comment (fwang) [inline packages/resources/src/Api.ts:207]: copy

File: packages/resources/src/Api.ts [hunk 5]
```diff
@@ -58,158 +74,183 @@ export interface RouteProps {
...
-      let routeProps = routes[routeName];
-      if (typeof routeProps === 'string') {
-        routeProps = { handler: routeProps };
+    // Validate input
+    if (!routes) {
+      throw new Error(`Missing 'routes' in sst.Api.`);
+    }
+    const routeKeys = Object.keys(routes);
+    if (routeKeys.length === 0) {
+      throw new Error("At least 1 route is required.");
```
Comment (fwang) [inline packages/resources/src/Api.ts:189]: copy

File: packages/resources/src/Api.ts [hunk 6]
```diff
@@ -58,158 +74,183 @@ export interface RouteProps {
...
     // Configure routes
     ///////////////////////////
 
-    Object.keys(routes).forEach((routeName: string) => {
-      let routeProps = routes[routeName];
-      if (typeof routeProps === 'string') {
-        routeProps = { handler: routeProps };
+    // Validate input
+    if (!routes) {
+      throw new Error(`Missing 'routes' in sst.Api.`);
```
Comment (fwang) [inline packages/resources/src/Api.ts:185]: copy

File: packages/resources/src/Api.ts [hunk 7]
```diff
@@ -58,158 +74,183 @@ export interface RouteProps {
...
+    apiProps.apiName = apiProps.apiName || `${Stack.of(this).stackName}-${id}`;
+    this.httpApi = new apig.HttpApi(this, "Api", apiProps);
 
     ///////////////////////////
     // Configure access log
     ///////////////////////////
 
+    // Validate input
+    if (accessLog !== undefined && httpApiProps !== undefined) {
+      throw new Error(`Cannot define both accessLog and httpApiProps.`);
```
Comment (fwang) [inline packages/resources/src/Api.ts:138]: copy

File: packages/resources/src/Api.ts [hunk 8]
```diff
@@ -58,158 +74,183 @@ export interface RouteProps {
...
-          apig.HttpMethod.OPTIONS,
-          apig.HttpMethod.PATCH,
-          apig.HttpMethod.POST,
-          apig.HttpMethod.PUT,
-        ],
-        allowOrigins: ['*'],
-      };
+    // Validate input
+    if (cors !== undefined && httpApiProps !== undefined) {
+      throw new Error(`Cannot define both cors and httpApiProps.`);
```
Comment (fwang) [inline packages/resources/src/Api.ts:108]: copy

**Review by jayair — APPROVED** [review]

**Pull request merged**